From dd5787b49a17a2e109ac8f342a2420d6039f5a07 Mon Sep 17 00:00:00 2001
From: Thomas Zeiser <5523252-tz-rrze@users.noreply.gitlab.com>
Date: Fri, 28 Feb 2020 17:38:09 +0100
Subject: [PATCH] fix detection of FFTW3 include file by configure

- If FFT_LIBS is set, it may nevertheless be necessary to use a
  special include path; thus, check FFT_INCLUDE and FFTW_INCLUDE.
  If they are set, add them to IFLAGS.
- If FFTW3 has to be autodetected, fix the test for fftw3.f03:
  + just 'include "fftw3.f03"' is not a valid Fortran program
    => 'use iso_c_binding // end' are missing
  + moreover, the file is in free form
    => '-ffree-form' as required as compile argument
  + the original 'for inc in $try_incdir' did not have any effect
    as the loop body did not make any use of $inc
    => only FFLAGS will be evaluated by AC_COMPILE_IFELSE

install/configure has to be updated, too. But that can easily be
done by running 'autoconf' and is therefore not included in the
commit.
---
 install/m4/x_ac_qe_fft.m4 | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/install/m4/x_ac_qe_fft.m4 b/install/m4/x_ac_qe_fft.m4
index a859ec9f9..1cc93d666 100644
--- a/install/m4/x_ac_qe_fft.m4
+++ b/install/m4/x_ac_qe_fft.m4
@@ -149,15 +149,20 @@ if test "$fft_libs" = ""; then
                   then
                         try_dflags="$try_dflags -D__FFTW3"
                         try_incdir="$FFTW_INCLUDE $FFTW_INC $INCLUDE_PATH $CPATH $FPATH"
+                        orig_fflags="$FFLAGS"
                         for inc in $try_incdir
                         do
-                           AC_COMPILE_IFELSE([include "fftw3.f03"],have_fft_include=1,)
+                           FFLAGS="$orig_fflags -I$inc -ffree-form"
+                           AC_COMPILE_IFELSE([use iso_c_binding
+include "fftw3.f03"
+end],have_fft_include=1,)
                            if test "$have_fft_include" -eq 1
                            then
                              try_iflags="$try_iflags -I$inc"
                              break
                            fi
                         done
+                        FFLAGS="$orig_fflags"
                         break
                   fi
 
@@ -174,6 +179,12 @@ if test "$fft_libs" = ""; then
 else
 
    echo "using FFT_LIBS with no testing ... "
+   if test -n "$FFT_INCLUDE" ; then :
+      try_iflags="$try_iflags -I$FFT_INCLUDE"
+   fi
+   if test -n "$FFTW_INCLUDE" ; then :
+      try_iflags="$try_iflags -I$FFTW_INCLUDE"
+   fi
 
 fi
 
-- 
GitLab
